packages/web-client/app/components/card-pay/create-merchant-workflow/merchant-customization/index.ts

```diff
@@ -0,0 +1,151 @@
+import { didCancel, restartableTask, timeout } from 'ember-concurrency';
+import { taskFor } from 'ember-concurrency-ts';
+import { mostReadable, random as randomColor } from '@ctrl/tinycolor';
+
+export default class CardPayDepositWorkflowTransactionAmountComponent extends Component<WorkflowCardComponentArgs> {
+  @service declare layer2Network: Layer2Network;
+  @tracked merchantBgColor: string = randomColor().toHexString();
+  @tracked merchantName: string = '';
+  @tracked merchantId: string = '';
+  @tracked lastCheckedMerchantId = '';
+  @tracked merchantNameValidationMessage = '';
+  @tracked merchantIdValidationMessage = '';
+  @tracked merchantBgColorValidationMessage = '';
+
+  get canSaveDetails() {
+    return (
+      // all fields populated
+      this.merchantBgColor &&
+      this.merchantName &&
+      this.merchantId &&
+      // unique merchant name + valid id characters
+      this.merchantIdInputState === 'valid' &&
+      // no other validation errors
+      !this.merchantNameValidationMessage &&
+      !this.merchantBgColorValidationMessage
+    );
+  }
+
+  get merchantIdInputState() {
+    if (taskFor(this.validateMerchantIdTask).isRunning) {
+      return 'loading';
+    } else if (
+      this.lastCheckedMerchantId === this.merchantId &&
+      taskFor(this.validateMerchantIdTask).last?.value
+    ) {
+      return 'valid';
+    } else if (this.merchantIdValidationMessage) {
+      return 'invalid';
+    } else {
+      return 'initial';
+    }
+  }
+
+  get merchantTextColor() {
+    return mostReadable(this.merchantBgColor, [
+      '#ffffff',
+      '#000000',
+    ])!.toHexString();
+  }
+
+  @action onMerchantNameInput(value: string) {
+    this.merchantName = value;
+    this.validateMerchantName();
+  }
+
+  @action onMerchantIdInput(value: string) {
+    this.merchantId = value;
+    this.validateMerchantId();
+  }
+
+  @action onMerchantBgColorInput(event: InputEvent) {
+    let value = (event.target as HTMLInputElement).value;
+    this.merchantBgColor = value;
+  }
+
+  @action saveDetails() {
+    this.args.workflowSession.updateMany({
+      merchantName: this.merchantName,
+      merchantId: this.merchantId,
+      merchantBgColor: this.merchantBgColor,
+      merchantTextColor: this.merchantTextColor,
+    });
+    this.args.onComplete?.();
+  }
+
+  @action validateMerchantName() {
+    if (this.merchantName) {
+      this.merchantNameValidationMessage = this.merchantName.trim()
+        ? ''
+        : 'Merchant name cannot contain only spaces';
+    } else {
+      this.merchantNameValidationMessage = 'This field is required';
+    }
+  }
+
+  @action async validateMerchantId() {
+    try {
+      await taskFor(this.validateMerchantIdTask).perform();
+    } catch (e) {
+      if (didCancel(e)) {
+        return;
+      }
+      console.error(e);
+    }
+  }
+
+  @restartableTask *validateMerchantIdTask() {
+    // debounce
+    yield timeout(500);
```

Comment:
I think this should be shorter in tests, ideally set in `config/environment` as with `threadAnimationInterval`, because it’s slowing the test suite unnecessarily:

![slow-test-validation-screencast 2021-08-20 10-11-43](https://user-images.githubusercontent.com/43280/130254807-556a2676-3f93-4ec7-9fb0-6ac74e081c8b.gif)

Here it is with 10 instead:

![fast-test-validation-screencast 2021-08-20 10-15-33](https://user-images.githubusercontent.com/43280/130255161-d44f2a46-4066-4b1c-a6b9-23a16a25c700.gif)

Perhaps the delays still seen in this faster one can also be eliminated, though it’s likely the 750 and therefore soon to be removed 🤔